tools/tf2openapi/generator/generate.go

```diff
@@ -1,15 +1,76 @@
 package generator
 
 import (
+	"fmt"
+	"github.com/getkin/kin-openapi/openapi3"
 	pb "github.com/kubeflow/kfserving/tools/tf2openapi/generated/protobuf"
+	"github.com/kubeflow/kfserving/tools/tf2openapi/types"
 )
 
 const defaultSigDefKey = "serving_default"
+const defaultTag = "serve"
 
-func GenerateOpenAPI(model *pb.SavedModel, sigDefKey string) (string, error) {
-	if sigDefKey == "" {
-		sigDefKey = defaultSigDefKey
+type Generator struct {
+	model         *pb.SavedModel
+	name          string
+	version       string
+	metaGraphTags []string
+	sigDefKey     string
+}
+
+func (g *Generator) WithName(name string) {
+	g.name = name
+}
+
+func (g *Generator) WithVersion(version string) {
+	g.version = version
+}
+
+func (g *Generator) WithMetaGraphTags(metaGraphTags []string) {
+	g.metaGraphTags = metaGraphTags
+}
+
+func (g *Generator) WithSigDefKey(sigDefKey string) {
+	g.sigDefKey = sigDefKey
+}
+
+func NewGenerator() Generator {
+	return Generator{
+		name:          "model",
+		version:       "1",
+		metaGraphTags: []string{defaultTag},
+		sigDefKey:     defaultSigDefKey,
+	}
+}
+
+func (g *Generator) GenerateOpenAPI(model *pb.SavedModel) (string, error) {
+	tfModel, constructionErr := types.NewTFSavedModel(model)
+	if constructionErr != nil {
+		return "", constructionErr
+	}
+	spec, genErr := g.tfServingOpenAPI(tfModel)
+	if genErr != nil {
+		return "", fmt.Errorf("missing info to generate OpenAPI specification\n error: %s", genErr.Error())
+	}
+	json, marshallingErr := (*spec).MarshalJSON()
```

Comment:
You shouldn't need to dereference spec to call MarshalJSON. What's going on here?